File: code/graph_algorithms/src/bellman_ford_algorithm/bellman_ford.js
/**
 * Part of Cosmos by OpenGenus
 * Javascript program for Bellman Ford algorithm.
 * Given a graph and a source vertex, it finds the
 * shortest path to all vertices from source vertex.
 */

class Graph {
  constructor(noOfVertices) {
    this.V = noOfVertices;
    this.graph = [];
  }

  addEdge(u, v, w) {
    this.graph.push([u, v, w]);
  }

  printSolution(dist) {
    console.log("Vertex Distance from Source");
    for (let v = 0; v < this.V; v++) {
      console.log(`${v} -> ${dist[v]}`);
    }
  }

  bellmanFord(src) {
    let dist = Array(this.V).fill(Infinity);
    dist[src] = 0;

    for (let i = 0; i < this.V - 1; i++) {
      for (const [u, v, w] of this.graph) {
        if (dist[u] != Infinity && dist[u] + w < dist[v]) {
          dist[v] = dist[u] + w;
        }
      }
    }

    // Check for negative weight cycle
    for (const [u, v, w] of this.graph) {
      if (dist[u] != Infinity && dist[u] + w < dist[v]) {
        console.log("Negative Weight Cycle is Present");
        return;
      }
    }

    this.printSolution(dist);
  }
}

g = new Graph(5);
g.addEdge(0, 1, -1);
g.addEdge(0, 2, 4);
g.addEdge(1, 2, 3);
g.addEdge(1, 3, 2);
g.addEdge(1, 4, 2);
g.addEdge(3, 2, 5);
g.addEdge(3, 1, 1);
g.addEdge(4, 3, -3);
g.bellmanFord(0);

File: code/Most_Useful_Algorithms/BellManFord.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

const int INF = INT_MAX; 


struct Edge {
    int src, dest, weight;
};

class Graph {
public:
    int V, E; 
    vector<Edge> edges;

    Graph(int vertices, int edges) : V(vertices), E(edges) {}

    
    void addEdge(int src, int dest, int weight) {
        Edge edge = {src, dest, weight};
        edges.push_back(edge);
    }


    void bellmanFord(int src) {
        vector<int> distance(V, INF);
        distance[src] = 0;

        
        for (int i = 1; i <= V - 1; i++) {
            for (const Edge& edge : edges) {
                int u = edge.src;
                int v = edge.dest;
                int weight = edge.weight;

                if (distance[u] != INF && distance[u] + weight < distance[v]) {
                    distance[v] = distance[u] + weight;
                }
            }
        }

    
        for (const Edge& edge : edges) {
            int u = edge.src;
            int v = edge.dest;
            int weight = edge.weight;

            if (distance[u] != INF && distance[u] + weight < distance[v]) {
                cout << "Graph contains negative-weight cycle." << endl;
                return;
            }
        }

        cout << "Shortest Distances from Source Vertex " << src << ":" << endl;
        for (int i = 0; i < V; i++) {
            cout << "Vertex " << i << ": ";
            if (distance[i] == INF) {
                cout << "INF" << endl;
            } else {
                cout << distance[i] << endl;
            }
        }
    }
};

int main() {
    int V, E;
    cout << "Enter the number of vertices (V) and edges (E): ";
    cin >> V >> E;

    Graph g(V, E);

    cout << "Enter the edges and their weights (format: src dest weight):" << endl;
    for (int i = 0; i < E; i++) {
        int src, dest, weight;
        cin >> src >> dest >> weight;
        g.addEdge(src, dest, weight);
    }

    int src;
    cout << "Enter the source vertex: ";
    cin >> src;

    g.bellmanFord(src);

    return 0;
}
